source/m2mnsdlinterface.cpp

```diff
@@ -1559,3 +1469,269 @@ nsdl_s * M2MNsdlInterface::get_nsdl_handle()
+    tr_debug("M2MNsdlInterface::handle_bootstrap_finished - path: %s", object_name.c_str());
+    sn_coap_hdr_s *coap_response = NULL;
+    uint8_t msg_code = COAP_MSG_CODE_RESPONSE_CHANGED;
+
+    // Accept only '/bs' path
+    if (object_name.size() != 2 ||
+            object_name.compare(0,2,BOOTSTRAP_URI) != 0 ||
+            !validate_security_object()) {
+        msg_code = COAP_MSG_CODE_RESPONSE_BAD_REQUEST;
+    } else {
+        // Add short server id to server object
+        _server->set_resource_value(M2MServer::ShortServerID,
+                                    _security->resource_value_int(M2MSecurity::ShortServerID));
+    }
+
+    coap_response = sn_nsdl_build_response(_nsdl_handle,
+                                           coap_header,
+                                           msg_code);
+    if(coap_response) {
+        sn_nsdl_send_coap_message(_nsdl_handle, address, coap_response);
+        sn_nsdl_release_allocated_coap_msg_mem(_nsdl_handle, coap_response);
+    }
+    if (COAP_MSG_CODE_RESPONSE_CHANGED == msg_code) {
+        _observer.bootstrap_done(_security);
+        // Switch back to original ep name
+        if (_endpoint->endpoint_name_ptr) {
+            memory_free(_endpoint->endpoint_name_ptr);
+        }
+        _endpoint->endpoint_name_ptr = alloc_string_copy((uint8_t*)_endpoint_name.c_str(), _endpoint_name.length());
+        _endpoint->endpoint_name_len = _endpoint_name.length();
+    } else {
+        handle_bootstrap_error();
+    }
+#else
+    (void) coap_header;
+    (void) address;
+#endif
+}
+
+void M2MNsdlInterface::handle_bootstrap_delete(sn_coap_hdr_s *coap_header,sn_nsdl_addr_s *address)
+{
+#ifndef M2M_CLIENT_DISABLE_BOOTSTRAP_FEATURE
+    tr_debug("M2MNsdlInterface::handle_bootstrap_delete");
+    sn_coap_hdr_s *coap_response = NULL;
+    coap_response = sn_nsdl_build_response(_nsdl_handle,
+                                           coap_header,
+                                           COAP_MSG_CODE_RESPONSE_DELETED);
+    if(coap_response) {
+        sn_nsdl_send_coap_message(_nsdl_handle, address, coap_response);
+        sn_nsdl_release_allocated_coap_msg_mem(_nsdl_handle, coap_response);
+        if(_security) {
+            _security->clear_resources();
+        }
+    } else {
+        handle_bootstrap_error();
+    }
+#else
+    (void) coap_header;
+    (void) address;
+#endif
+}
+
+bool M2MNsdlInterface::validate_security_object()
+{
+#ifndef M2M_CLIENT_DISABLE_BOOTSTRAP_FEATURE
+    tr_debug("M2MNsdlInterface::validate_security_object");
+    if (_security) {
+        String address = _security->resource_value_string(M2MSecurity::M2MServerUri);
+        uint32_t sec_mode = _security->resource_value_int(M2MSecurity::SecurityMode);
+        bool is_bs_server = _security->resource_value_int(M2MSecurity::BootstrapServer);
+        uint8_t *buffer = NULL;
+        uint32_t public_key_size = _security->resource_value_buffer(M2MSecurity::PublicKey, buffer);
```

Comment:
You can use value_length() to get the size instead of populating the buffer each time.